ideascube/serveradmin/models.py

```diff
@@ -0,0 +1,27 @@
+from django.conf import settings
+from django.db import models
+
+from ideascube.models import JSONField
+
+
+class Setting(models.Model):
+    class Meta:
+        ordering = ['-date']
+
+    namespace = models.CharField(max_length=40)
+    key = models.CharField(max_length=40)
+    value = JSONField()
+    actor = models.ForeignKey(settings.AUTH_USER_MODEL)
+    date = models.DateTimeField(auto_now=True)
+
+    def as_boolean(self):
```

Comment:
So at first here, I had those methods that weren't casting.

Instead, they were ensuring that the value was of the given type, and raising `TypeError` if not.

That is, I had:

```python
    def as_boolean(self):
        if not isinstance(self, bool):
            raise TypeError('%s is not a boolean' % self)

        return self.value

    def as_integer(self):
        if not isinstance(self, int):
            raise TypeError('%s is not an integer' % self)

        return self.value

    def as_string(self):
        if not isinstance(self, str):
            raise TypeError('%s is not a string' % self)

        return self.value
```

I found it more interesting, because it served as validation of the setting values, and I was happy. The `default` parameters you suggest would have made **lots** of sense here.

Then went on to implement the tests, and was surprised when they didn't do what I expect.

I was expecting that a string setting to raise for all the `as_*` methods but `as_string`, an integer setting to raise for all the `as_*` method but `as_integer`, etc...

But what happened is that a boolean setting raised only for `as_string`, it successfully passed through for `as_boolean` (of course) and `as_integer`. Because of course, `bool` is a subclass of `int`!

That made me realize that the validation provided by these methods wasn't as good as I initially thought.

Then I started questioning whether it even was useful in the first place.

Then I decided that instead of validation, I'd just cast, which is another way of ensuring I'm getting a value in the right type (although it might be a completely bogus value).

But to be honest, I'm now less convinced that these methods are useful at all.

And I don't think the `default` you suggest makes sense in this case, either.

---

Fast forward to today, I've slept on it, and I'm glad you didn't merge it. :smile: 

This morning I'm even less convinced by those `as_*` methods than I was yesterday.

I mean, they ensure that if I want a string setting I'm going to get a string setting, sure.

But I still have to do something like:

```python
try:
    setting = Setting.objects.get(namespace='foo', key='bar')

except Setting.DoesNotExist:
    value = default_value

else:
    try:
        value = setting.as_<something>()

    except TypeError:
        value = default_value
```

That's a pretty horrendous API!

With the `default` you suggest, it gets better:

```python
try:
    setting = Setting.objects.get(namespace='foo', key='bar')

except Setting.DoesNotExist:
    value = default_value

else:
    value = setting.as_<something>(default=default_value)
```

But it's still pretty bad API. :smiley: 

This would be a much better API:

```python
value = Setting.get_<something>('foo', 'bar', default=default_value)
```

That would mean making those current `as_*` methods into `get_*` class methods, implemented something like this:

```python
    @classmethod
    def get_string(cls, namespace, key, default_value=None):
        try:
            setting = Setting.objects.get(namespace=namespace, key=key)

        except Setting.DoesNotExist:
            if default_value is not None:
                return default_value

            raise

        else:
            if insinstance(setting.value, str):
                return setting.value

            if default_value is not None:
                return default_value

            raise TypeError('%s is not a string setting' % setting)
```

Yes, I went back to checking the type instead of casting. I think this is better, and the problem I had with boolean settings is... well I don't really think it's a problem after a good night's sleep.

Of course, rather than class methods, maybe manager methods would be more idiomatic Django, even though they'd make for worse API:

```python
value = Setting.objects.get_<something>('foo', 'bar', default=default_value)
```

(and for the record, this is what I had actually started with yesterday, but then changed my mind)

What do you think?